### PR TITLE
Generic Assay plots tab implementation(Categorical / Binary)

### PR DESCRIPTION
**Pull request by dippindots**
Fix cBioPortal/cbioportal#7935

**Feature:**
General plots tab implementation of Categorical / Binary data in Generic Assay format.

**Example:**
Arm-level Cna data
image

**Test study:** 
http://private.cbioportal.org/private/study/summary?id=acc_tcga_pan_can_atlas_2018

**Test instruction:** 
1. Go to study view page: http://private.cbioportal.org/private/study/summary?id=acc_tcga_pan_can_atlas_2018
2. Set `localStorage.netlify ="deploy-preview-3694--cbioportalfrontend"` in browser console.
3. Query any gene, and go to result view page
4. Select plots tab
5. Select `Armlevel Cna` in `Data Type`

**Please note: this is not the final feature for `Armlevel Cna` data, we may do further customization for `Armlevel Cna` data, currently, it's been used as a test dataset.**

### CODE REVIEW

File: src/shared/cache/GenericAssayMolecularDataCache.ts [hunk 1]
```diff
@@ -60,7 +60,7 @@ function handleValueThreshold(
 ): GenericAssayDataEnhanced {
     if (!datum.thresholdType) {
         // this is to prevent repeated extraction of '>' and '<' symbols
-        const matches = /([><]?)(.+)/.exec(datum.value);
+        const matches = /([><]?)(-?\\d+(\\.\\d+)?)/.exec(datum.value);
```
Comment (dippindots) [inline src/shared/cache/GenericAssayMolecularDataCache.ts:63]: Previous code will match any string regardless it's numeric or not. That was correct at that moment, but we have free text data now. So we need to change this to match numeric data only.

**Comment by dippindots** [comment]
@cBioPortal/product The general implementation for Categorical / Binary Generic Assay data is ready for test, here is the instruction of how to test this feature: https://github.com/cBioPortal/cbioportal-frontend/pull/3694#issue-603272944. Thanks for your time!

File: src/pages/resultsView/plots/PlotsTabUtils.tsx [hunk 2]
```diff
@@ -2374,7 +2380,10 @@ export function logScalePossibleForProfile(profileId: string) {
 
 export function logScalePossible(axisSelection: AxisMenuSelection) {
     if (axisSelection.dataType !== CLIN_ATTR_DATA_TYPE) {
-        if (axisSelection.dataType && axisSelection.isGenericAssayType) {
+        if (
+            isGenericAssaySelected(axisSelection) &&
```
Comment (alisman) [inline src/pages/resultsView/plots/PlotsTabUtils.tsx:2384]: why do we bother checking above for CLIN_ATTR_DATA_TYPE?    we don't want to make log scale availabe for any of this?

Comment (alisman) [reply src/pages/resultsView/plots/PlotsTabUtils.tsx:2384]: maybe put a a comment explaining this rule

File: src/shared/cache/GenericAssayMolecularDataCache.ts [hunk 3]
```diff
@@ -59,16 +59,18 @@ function handleValueThreshold(
     datum: GenericAssayDataEnhanced
 ): GenericAssayDataEnhanced {
     if (!datum.thresholdType) {
-        // this is to prevent repeated extraction of '>' and '<' symbols
-        const matches = /([><]?)(.+)/.exec(datum.value);
+        // only extract threshold for numeric values
```
Comment (alisman) [inline src/shared/cache/GenericAssayMolecularDataCache.ts:62]: can you add a comment explaining what threshold means here?

File: src/shared/cache/GenericAssayMolecularDataCache.ts [hunk 4]
```diff
@@ -59,16 +59,18 @@ function handleValueThreshold(
...
+        const matches = /^([><]? *)(-?\d+(\.\d+)?)$/.exec(datum.value);
```
Comment (alisman) [inline src/shared/cache/GenericAssayMolecularDataCache.ts:63]: add an comment with example of what this regexp matches

**Review by alisman — APPROVED** [review]

**Comment by jjgao** [comment]
- [ ] rename "Armlevel Cna" to "Arm-level CNA" (data)
- [ ] Change the order in Plots to (lost, gained, diploid, NA)
- [ ] Do we need to keep NA in the chart? (we should not have NA in the database)
image
- [ ] adding charts based for "Arm-level CNA" is not working.

**Comment by dippindots** [comment]
@jjgao Thanks for your review! The second one `` is not showing because I haven't finished the study view implementation, I created a ticket [here](https://github.com/cBioPortal/cbioportal/issues/8557) to disable unfinished features for now.

**Comment by dippindots** [comment]
After rebased and checked screenshots, all looks good, merging this into feature branch `categorical-generic-assay`

**Pull request merged**